File: internal/daemon/worktree.go
package daemon

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type WorktreeManager struct {
	RepoRoot string
}

func NewWorktreeManager(repoRoot string) *WorktreeManager {
	return &WorktreeManager{RepoRoot: repoRoot}
}

func ValidateGitRepo(dir string) (string, error) {
	return gitOutput(dir, "rev-parse", "--show-toplevel")
}

func CurrentBranch(dir string) string {
	out, err := gitOutput(dir, "branch", "--show-current")
	if err != nil || out == "" {
		return "detached"
	}
	return out
}

func (w *WorktreeManager) EnsureGitignore() error {
	gitignorePath := filepath.Join(w.RepoRoot, ".gitignore")
	excludePath := filepath.Join(w.RepoRoot, ".git", "info", "exclude")
	target := gitignorePath
	if _, err := os.Stat(gitignorePath); err != nil {
		target = excludePath
	}
	content, _ := os.ReadFile(target)
	if bytes.Contains(content, []byte(".spawntree/")) {
		return nil
	}
	f, err := os.OpenFile(target, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.WriteString("\n.spawntree/\n")
	return err
}

func (w *WorktreeManager) Create(envName string) (string, error) {
	path := filepath.Join(w.RepoRoot, ".spawntree", "envs", envName)
	if _, err := os.Stat(path); err == nil {
		return path, nil
	}
	cmd := exec.Command("git", "worktree", "add", path, "HEAD", "--detach")
	cmd.Dir = w.RepoRoot
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("failed to create worktree for %q: %s", envName, strings.TrimSpace(string(output)))
	}
	return path, nil
}

func (w *WorktreeManager) Remove(envName string) error {
	path := filepath.Join(w.RepoRoot, ".spawntree", "envs", envName)
	if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
		return nil
	} else if err != nil {
		return err
	}
	cmd := exec.Command("git", "worktree", "remove", path, "--force")
	cmd.Dir = w.RepoRoot
	if output, err := cmd.CombinedOutput(); err != nil {
		_ = os.RemoveAll(path)
		prune := exec.Command("git", "worktree", "prune")
		prune.Dir = w.RepoRoot
		_, _ = prune.CombinedOutput()
		if len(output) > 0 {
			return errors.New(strings.TrimSpace(string(output)))
		}
	}
	return nil
}

// GitWorktreeInfo holds parsed output from `git worktree list --porcelain`.
type GitWorktreeInfo struct {
	Path    string
	Branch  string
	HeadRef string
}

// listGitWorktrees runs `git worktree list --porcelain` and parses the output.
func listGitWorktrees(dir string) ([]GitWorktreeInfo, error) {
	out, err := gitOutput(dir, "worktree", "list", "--porcelain")
	if err != nil {
		return nil, err
	}
	if out == "" {
		return nil, nil
	}

	var worktrees []GitWorktreeInfo
	var current GitWorktreeInfo

	for _, line := range strings.Split(out, "\n") {
		switch {
		case strings.HasPrefix(line, "worktree "):
			if current.Path != "" {
				worktrees = append(worktrees, current)
			}
			current = GitWorktreeInfo{Path: strings.TrimPrefix(line, "worktree ")}
		case strings.HasPrefix(line, "HEAD "):
			current.HeadRef = strings.TrimPrefix(line, "HEAD ")
		case strings.HasPrefix(line, "branch "):
			ref := strings.TrimPrefix(line, "branch ")
			// Strip refs/heads/ prefix
			current.Branch = strings.TrimPrefix(ref, "refs/heads/")
		case line == "detached":
			current.Branch = ""
		}
	}
	if current.Path != "" {
		worktrees = append(worktrees, current)
	}

	return worktrees, nil
}

func gitOutput(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", errors.New(strings.TrimSpace(string(output)))
	}
	return strings.TrimSpace(string(output)), nil
}

File: packages/core/src/env/worktree.ts
import { execSync } from "node:child_process";
import { appendFileSync, existsSync, readFileSync, rmSync } from "node:fs";
import { resolve } from "node:path";

export class WorktreeManager {
  private readonly repoRoot: string;
  private readonly spawntreeDir: string;

  constructor(repoRoot: string) {
    this.repoRoot = repoRoot;
    this.spawntreeDir = resolve(repoRoot, ".spawntree");
  }

  /**
   * Validate that we're inside a git repository.
   * Throws with a clear error if not.
   */
  static validateGitRepo(dir: string): string {
    try {
      const root = execSync("git rev-parse --show-toplevel", {
        cwd: dir,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
      }).trim();
      return root;
    } catch {
      throw new Error(
        `Not a git repository: ${dir}\nspawntree requires a git repository for environment isolation.`,
      );
    }
  }

  /**
   * Get the current git branch name.
   */
  static currentBranch(dir: string): string {
    try {
      return execSync("git branch --show-current", {
        cwd: dir,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
      }).trim();
    } catch {
      return "detached";
    }
  }

  /**
   * Ensure .spawntree/ is in .gitignore.
   */
  ensureGitignore(): void {
    const gitignorePath = resolve(this.repoRoot, ".gitignore");
    const excludePath = resolve(this.repoRoot, ".git", "info", "exclude");

    if (existsSync(gitignorePath)) {
      const content = readFileSync(gitignorePath, "utf-8");
      if (!content.includes(".spawntree/")) {
        appendFileSync(gitignorePath, "\n.spawntree/\n");
      }
    } else {
      appendFileSync(excludePath, "\n.spawntree/\n");
    }
  }

  /**
   * Create a git worktree for the given environment.
   * Returns the path to the worktree directory.
   */
  create(envName: string): string {
    const worktreePath = resolve(this.spawntreeDir, "envs", envName);

    if (existsSync(worktreePath)) {
      return worktreePath;
    }

    try {
      execSync(
        `git worktree add "${worktreePath}" HEAD --detach`,
        {
          cwd: this.repoRoot,
          encoding: "utf-8",
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      throw new Error(`Failed to create worktree for "${envName}": ${msg}`, { cause: err });
    }

    return worktreePath;
  }

  /**
   * Remove a git worktree for the given environment.
   */
  remove(envName: string): void {
    const worktreePath = resolve(this.spawntreeDir, "envs", envName);

    if (!existsSync(worktreePath)) return;

    try {
      execSync(`git worktree remove "${worktreePath}" --force`, {
        cwd: this.repoRoot,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
      });
    } catch {
      // Fallback: just remove the directory
      rmSync(worktreePath, { recursive: true, force: true });
      try {
        execSync("git worktree prune", {
          cwd: this.repoRoot,
          stdio: ["pipe", "pipe", "pipe"],
        });
      } catch {
        // ignore
      }
    }
  }

  /**
   * Check if a worktree exists for the given environment.
   */
  exists(envName: string): boolean {
    return existsSync(resolve(this.spawntreeDir, "envs", envName));
  }

  /**
   * Get the worktree path for the given environment.
   */
  path(envName: string): string {
    return resolve(this.spawntreeDir, "envs", envName);
  }
}
